modules/benchmark/src/main/scala/io/circe/benchmark/FoldingBenchmark.scala

```diff
@@ -0,0 +1,69 @@
+package io.circe.benchmark
+
+import io.circe.{ Json, JsonNumber, JsonObject }
+import java.util.concurrent.TimeUnit
+import org.openjdk.jmh.annotations._
+
+/**
+ * Compare the performance of various ways of folding JSON values.
+ *
+ * The following command will run the benchmarks with reasonable settings:
+ *
+ * > sbt "benchmark/jmh:run -i 10 -wi 10 -f 2 -t 1 io.circe.benchmark.FoldingBenchmark"
+ */
+@State(Scope.Thread)
+@BenchmarkMode(Array(Mode.Throughput))
+@OutputTimeUnit(TimeUnit.SECONDS)
+class FoldingBenchmark extends ExampleData {
+  val doc: Json = Json.arr(intsJson, booleansJson, foosJson, helloWorldJson)
+
+  @Benchmark
+  def withFoldWith: Int = doc.foldWith(
+    new Json.Folder[Int] {
```

Comment:
:+1:, yeah, this is neat. It only has a tiny effect on allocations in this benchmark, but I think it looks a little nicer anyway, so let's consider it idiomatic.